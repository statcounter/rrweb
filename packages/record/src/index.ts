import {
  record,
  getWindowScroll,
  getWindowHeight,
  getWindowWidth,
} from 'rrweb';

import { EventType, IncrementalSource, MouseInteractions } from '@rrweb/types';

export {
  EventType,
  IncrementalSource,
  MouseInteractions,
  record,
  getWindowScroll,
  getWindowHeight,
  getWindowWidth,
};
