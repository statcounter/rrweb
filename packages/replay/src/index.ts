import {
  Replayer,
  type playerConfig,
  type PlayerMachineState,
  type SpeedMachineState,
} from 'rrweb';
import 'rrweb/dist/style.css';

import {
  EventType,
  IncrementalSource,
  MouseInteractions,
} from '@rrweb/types';

export {
  Replayer,
  EventType,
  IncrementalSource,
  MouseInteractions,
  type playerConfig,
  type PlayerMachineState,
  type SpeedMachineState,
};
