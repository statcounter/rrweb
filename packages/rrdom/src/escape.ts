// for serializing a virtual DOM to html

export function escapeHtmlText(text: string): string {
  return text
    .replace(/&/g, '&amp;')
    .replace(/ /g, '&nbsp;') // unicode version is correct, but output the html escape syntax for consistency with what chrome outputs
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;');
}

export function escapeHtmlAttr(text: string): string {
  return text
    .replace(/&/g, '&amp;')
    .replace(/ /g, '&nbsp;') // unicode version is correct, but output the html escape syntax for consistency with what chrome outputs
    .replace(/"/g, '&quot;');
}
