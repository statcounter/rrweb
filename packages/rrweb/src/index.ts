import record from './record';
import {
  Replayer,
  type playerConfig,
  type PlayerMachineState,
  type SpeedMachineState,
} from './replay';
import { SyncReplayer } from './replay/sync-replayer';

import canvasMutation from './replay/canvas';
import { _mirror } from './utils';
import * as utils from './utils';

export {
  EventType,
  IncrementalSource,
  MouseInteractions,
  ReplayerEvents,
  type eventWithTime,
} from '@rrweb/types';

// exports style.css from replay
import './replay/styles/style.css';

export type { recordOptions, ReplayPlugin } from './types';

const { addCustomEvent } = record;
const { freezePage } = record;
const { takeFullSnapshot } = record;

const getWindowScroll = utils.getWindowScroll;
const getWindowHeight = utils.getWindowHeight;
const getWindowWidth = utils.getWindowWidth;

export {
  record,
  addCustomEvent,
  freezePage,
  takeFullSnapshot,
  Replayer,
  SyncReplayer,
  type playerConfig,
  type PlayerMachineState,
  type SpeedMachineState,
  canvasMutation,
  _mirror as mirror,
  utils,
  getWindowScroll,
  getWindowHeight,
  getWindowWidth,
};
