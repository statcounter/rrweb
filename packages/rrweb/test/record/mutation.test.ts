/**
 * @vitest-environment jsdom
 */
//import MutationBuffer from '../mutation';
//import type { MutationBufferParam } from '../../types';
/*

describe('Processing mutations', () => {

  describe('Large list 1 perf', () => {
    const mutationBuffer = new MutationBuffer();
    const options: MutationBufferParam = {
    };
    mutationBuffer.init(options);

    mutationBuffer.processMutations([]);
  });
*/

import * as fs from 'fs';
import * as path from 'path';
import { vi } from 'vitest';

import {
  assertSnapshot,
  getServerURL,
  ISuite,
  launchPuppeteer,
  startServer,
  waitForRAF,
} from '../utils';
import { EventType, eventWithTime, listenerHandler } from '@rrweb/types';
import { recordOptions } from '../../src/types';

interface IWindow extends Window {
  rrweb: {
    record: (
      options: recordOptions<eventWithTime>,
    ) => listenerHandler | undefined;
    addCustomEvent<T>(tag: string, payload: T): void;
  };
  emit: (e: eventWithTime) => undefined;
}

describe('mutation', () => {
  vi.setConfig({ testTimeout: 100_000 });
  let code: ISuite['code'];
  let page: ISuite['page'];
  let browser: ISuite['browser'];
  let server: ISuite['server'];
  let serverURL: ISuite['serverURL'];
  let events: ISuite['events'];

  beforeAll(async () => {
    server = await startServer();
    serverURL = getServerURL(server);
    browser = await launchPuppeteer();

    const bundlePath = path.resolve(__dirname, '../../dist/rrweb.umd.cjs');
    code = fs.readFileSync(bundlePath, 'utf8');
  });

  afterEach(async () => {
    await page.close();
  });

  afterAll(async () => {
    await server.close();
    await browser.close();
  });

  beforeEach(async () => {
    page = await browser.newPage();
    page.on('console', (msg) => {
      console.log(msg.text());
    });

    await page.goto(`${serverURL}/html/mutation.html`);
    await page.addScriptTag({
      path: path.resolve(__dirname, '../../dist/rrweb.umd.cjs'),
    });
    await waitForRAF(page);
    events = [];

    await page.exposeFunction('emit', (e: eventWithTime) => {
      if (e.type === EventType.DomContentLoaded || e.type === EventType.Load) {
        return;
      }
      events.push(e);
    });

    page.on('console', (msg) => console.log('PAGE LOG:', msg.text()));

    await page.evaluate(() => {
      const { record } = (window as unknown as IWindow).rrweb;
      record({
        emit: (window as unknown as IWindow).emit,
        slimDOMOptions: {
          comment: true,
        },
      });
    });

    await waitForRAF(page);
  });

  it('add elements all at once', async () => {
    await page.evaluate(() => {
      const d1 = document.createElement('div');
      d1.id = 'd1';
      const s1 = document.createElement('span');
      d1.append(s1);
      const d2 = document.createElement('div');
      d2.id = 'd2';
      const s2 = document.createElement('span');
      d2.append(s2);
      const d3 = document.createElement('div');
      d3.id = 'd3';
      const s3 = document.createElement('span');
      d3.append(s3);

      d1.append(d2);
      d1.append(d3);
      document.body.append(d1);
    });
    await waitForRAF(page);
    const mutations = events.filter(
      (e) => e.type === EventType.IncrementalSnapshot,
    );
    await assertSnapshot(mutations, true);
  });

  it('add root first', async () => {
    await page.evaluate(() => {
      const d1 = document.createElement('div');
      d1.id = 'd1';
      const s1 = document.createElement('span');
      d1.append(s1);
      document.body.append(d1);
      const d2 = document.createElement('div');
      d2.id = 'd2';
      const s2 = document.createElement('span');
      d2.append(s2);
      const d3 = document.createElement('div');
      d3.id = 'd3';
      const s3 = document.createElement('span');
      d3.append(s3);

      d1.append(d2);
      d1.append(d3);
    });
    await waitForRAF(page);
    const mutations = events.filter(
      (e) => e.type === EventType.IncrementalSnapshot,
    );

    // assert has same output as previous test despite difference in way elements are added to DOM
    //await assertSnapshot(mutations, 'mutation.test.ts.mutation.add_elements_all_at_once');

    await assertSnapshot(mutations, true);
  });

  it('ignored firstchild comment', async () => {
    await page.evaluate(() => {
      const d1 = document.createElement('div');
      document.body.append(d1);
      const c1 = document.createComment('slimdom ignored');
      const siblingDiv = document.createElement('div');
      d1.append(c1);
      d1.append(siblingDiv);
    });
    await waitForRAF(page);
    const mutations = events.filter(
      (e) => e.type === EventType.IncrementalSnapshot,
    );
    await assertSnapshot(mutations, true);
  });

  it('ignored middlechild comment', async () => {
    await page.evaluate(() => {
      const d1 = document.createElement('div');
      document.body.append(d1);
      const c1 = document.createComment('slimdom ignored');
      const siblingDiv = document.createElement('div');
      const siblingDiv2 = document.createElement('div');
      d1.append(siblingDiv);
      d1.append(c1);
      d1.append(siblingDiv2);
    });
    await waitForRAF(page);
    const mutations = events.filter(
      (e) => e.type === EventType.IncrementalSnapshot,
    );
    await assertSnapshot(mutations, true);
  });

  it('ignored endchild comment', async () => {
    await page.evaluate(() => {
      const d1 = document.createElement('div');
      document.body.append(d1);
      const c1 = document.createComment('slimdom ignored');
      const siblingDiv = document.createElement('div');
      d1.append(siblingDiv);
      d1.append(c1);
    });
    await waitForRAF(page);
    const mutations = events.filter(
      (e) => e.type === EventType.IncrementalSnapshot,
    );
    await assertSnapshot(mutations, true);
  });
});
