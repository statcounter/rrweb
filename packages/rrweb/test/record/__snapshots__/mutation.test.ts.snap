// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mutation > add elements at once 1`] = `
"[
  {
    \\"type\\": 4,
    \\"data\\": {
      \\"href\\": \\"about:blank\\",
      \\"width\\": 1920,
      \\"height\\": 1080
    }
  },
  {
    \\"type\\": 2,
    \\"data\\": {
      \\"node\\": {
        \\"type\\": 0,
        \\"childNodes\\": [
          {
            \\"type\\": 1,
            \\"name\\": \\"html\\",
            \\"publicId\\": \\"\\",
            \\"systemId\\": \\"\\",
            \\"id\\": 2
          },
          {
            \\"type\\": 2,
            \\"tagName\\": \\"html\\",
            \\"attributes\\": {},
            \\"childNodes\\": [
              {
                \\"type\\": 2,
                \\"tagName\\": \\"head\\",
                \\"attributes\\": {},
                \\"childNodes\\": [
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"script\\",
                    \\"attributes\\": {
                      \\"type\\": \\"text/javascript\\"
                    },
                    \\"childNodes\\": [
                      {
                        \\"type\\": 3,
                        \\"textContent\\": \\"SCRIPT_PLACEHOLDER\\",
                        \\"id\\": 6
                      }
                    ],
                    \\"id\\": 5
                  }
                ],
                \\"id\\": 4
              },
              {
                \\"type\\": 2,
                \\"tagName\\": \\"body\\",
                \\"attributes\\": {},
                \\"childNodes\\": [
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n  \\\\n\\\\n\\",
                    \\"id\\": 8
                  }
                ],
                \\"id\\": 7
              }
            ],
            \\"id\\": 3
          }
        ],
        \\"id\\": 1
      },
      \\"initialOffset\\": {
        \\"left\\": 0,
        \\"top\\": 0
      }
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 0,
      \\"texts\\": [],
      \\"attributes\\": [],
      \\"removes\\": [],
      \\"adds\\": [
        {
          \\"parentId\\": 7,
          \\"nextId\\": null,
          \\"node\\": {
            \\"type\\": 2,
            \\"tagName\\": \\"div\\",
            \\"attributes\\": {
              \\"id\\": \\"d1\\"
            },
            \\"childNodes\\": [],
            \\"id\\": 9
          }
        },
        {
          \\"parentId\\": 9,
          \\"nextId\\": null,
          \\"node\\": {
            \\"type\\": 2,
            \\"tagName\\": \\"div\\",
            \\"attributes\\": {
              \\"id\\": \\"d3\\"
            },
            \\"childNodes\\": [],
            \\"id\\": 10
          }
        },
        {
          \\"parentId\\": 9,
          \\"nextId\\": 10,
          \\"node\\": {
            \\"type\\": 2,
            \\"tagName\\": \\"div\\",
            \\"attributes\\": {
              \\"id\\": \\"d2\\"
            },
            \\"childNodes\\": [],
            \\"id\\": 11
          }
        },
        {
          \\"parentId\\": 9,
          \\"nextId\\": 11,
          \\"node\\": {
            \\"type\\": 2,
            \\"tagName\\": \\"span\\",
            \\"attributes\\": {},
            \\"childNodes\\": [],
            \\"id\\": 12
          }
        },
        {
          \\"parentId\\": 11,
          \\"nextId\\": null,
          \\"node\\": {
            \\"type\\": 2,
            \\"tagName\\": \\"span\\",
            \\"attributes\\": {},
            \\"childNodes\\": [],
            \\"id\\": 13
          }
        },
        {
          \\"parentId\\": 10,
          \\"nextId\\": null,
          \\"node\\": {
            \\"type\\": 2,
            \\"tagName\\": \\"span\\",
            \\"attributes\\": {},
            \\"childNodes\\": [],
            \\"id\\": 14
          }
        }
      ]
    }
  }
]"
`;

exports[`mutation > add root first 1`] = `
"[
  {
    \\"type\\": 4,
    \\"data\\": {
      \\"href\\": \\"about:blank\\",
      \\"width\\": 1920,
      \\"height\\": 1080
    }
  },
  {
    \\"type\\": 2,
    \\"data\\": {
      \\"node\\": {
        \\"type\\": 0,
        \\"childNodes\\": [
          {
            \\"type\\": 1,
            \\"name\\": \\"html\\",
            \\"publicId\\": \\"\\",
            \\"systemId\\": \\"\\",
            \\"id\\": 2
          },
          {
            \\"type\\": 2,
            \\"tagName\\": \\"html\\",
            \\"attributes\\": {},
            \\"childNodes\\": [
              {
                \\"type\\": 2,
                \\"tagName\\": \\"head\\",
                \\"attributes\\": {},
                \\"childNodes\\": [
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"script\\",
                    \\"attributes\\": {
                      \\"type\\": \\"text/javascript\\"
                    },
                    \\"childNodes\\": [
                      {
                        \\"type\\": 3,
                        \\"textContent\\": \\"SCRIPT_PLACEHOLDER\\",
                        \\"id\\": 6
                      }
                    ],
                    \\"id\\": 5
                  }
                ],
                \\"id\\": 4
              },
              {
                \\"type\\": 2,
                \\"tagName\\": \\"body\\",
                \\"attributes\\": {},
                \\"childNodes\\": [
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n  \\\\n\\\\n\\",
                    \\"id\\": 8
                  }
                ],
                \\"id\\": 7
              }
            ],
            \\"id\\": 3
          }
        ],
        \\"id\\": 1
      },
      \\"initialOffset\\": {
        \\"left\\": 0,
        \\"top\\": 0
      }
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 0,
      \\"texts\\": [],
      \\"attributes\\": [],
      \\"removes\\": [],
      \\"adds\\": [
        {
          \\"parentId\\": 7,
          \\"nextId\\": null,
          \\"node\\": {
            \\"type\\": 2,
            \\"tagName\\": \\"div\\",
            \\"attributes\\": {
              \\"id\\": \\"d1\\"
            },
            \\"childNodes\\": [],
            \\"id\\": 9
          }
        },
        {
          \\"parentId\\": 9,
          \\"nextId\\": null,
          \\"node\\": {
            \\"type\\": 2,
            \\"tagName\\": \\"div\\",
            \\"attributes\\": {},
            \\"childNodes\\": [],
            \\"id\\": 10
          }
        },
        {
          \\"parentId\\": 9,
          \\"nextId\\": 10,
          \\"node\\": {
            \\"type\\": 2,
            \\"tagName\\": \\"div\\",
            \\"attributes\\": {
              \\"id\\": \\"d2\\"
            },
            \\"childNodes\\": [],
            \\"id\\": 11
          }
        },
        {
          \\"parentId\\": 9,
          \\"nextId\\": 11,
          \\"node\\": {
            \\"type\\": 2,
            \\"tagName\\": \\"span\\",
            \\"attributes\\": {},
            \\"childNodes\\": [],
            \\"id\\": 12
          }
        },
        {
          \\"parentId\\": 11,
          \\"nextId\\": null,
          \\"node\\": {
            \\"type\\": 2,
            \\"tagName\\": \\"span\\",
            \\"attributes\\": {},
            \\"childNodes\\": [],
            \\"id\\": 13
          }
        },
        {
          \\"parentId\\": 10,
          \\"nextId\\": null,
          \\"node\\": {
            \\"type\\": 2,
            \\"tagName\\": \\"span\\",
            \\"attributes\\": {},
            \\"childNodes\\": [],
            \\"id\\": 14
          }
        }
      ]
    }
  }
]"
`;

exports[`mutation > ignored ending comment 1`] = `
"[
  {
    \\"type\\": 4,
    \\"data\\": {
      \\"href\\": \\"about:blank\\",
      \\"width\\": 1920,
      \\"height\\": 1080
    }
  },
  {
    \\"type\\": 2,
    \\"data\\": {
      \\"node\\": {
        \\"type\\": 0,
        \\"childNodes\\": [
          {
            \\"type\\": 1,
            \\"name\\": \\"html\\",
            \\"publicId\\": \\"\\",
            \\"systemId\\": \\"\\",
            \\"id\\": 2
          },
          {
            \\"type\\": 2,
            \\"tagName\\": \\"html\\",
            \\"attributes\\": {},
            \\"childNodes\\": [
              {
                \\"type\\": 2,
                \\"tagName\\": \\"head\\",
                \\"attributes\\": {},
                \\"childNodes\\": [
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"script\\",
                    \\"attributes\\": {
                      \\"type\\": \\"text/javascript\\"
                    },
                    \\"childNodes\\": [
                      {
                        \\"type\\": 3,
                        \\"textContent\\": \\"SCRIPT_PLACEHOLDER\\",
                        \\"id\\": 6
                      }
                    ],
                    \\"id\\": 5
                  }
                ],
                \\"id\\": 4
              },
              {
                \\"type\\": 2,
                \\"tagName\\": \\"body\\",
                \\"attributes\\": {},
                \\"childNodes\\": [
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n  \\\\n\\\\n\\",
                    \\"id\\": 8
                  }
                ],
                \\"id\\": 7
              }
            ],
            \\"id\\": 3
          }
        ],
        \\"id\\": 1
      },
      \\"initialOffset\\": {
        \\"left\\": 0,
        \\"top\\": 0
      }
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 0,
      \\"texts\\": [],
      \\"attributes\\": [],
      \\"removes\\": [],
      \\"adds\\": [
        {
          \\"parentId\\": 7,
          \\"nextId\\": null,
          \\"node\\": {
            \\"type\\": 2,
            \\"tagName\\": \\"div\\",
            \\"attributes\\": {},
            \\"childNodes\\": [],
            \\"id\\": 9
          }
        },
        {
          \\"parentId\\": 9,
          \\"nextId\\": null,
          \\"node\\": {
            \\"type\\": 2,
            \\"tagName\\": \\"div\\",
            \\"attributes\\": {},
            \\"childNodes\\": [],
            \\"id\\": 10
          }
        }
      ]
    }
  }
]"
`;

exports[`mutation > ignored middle comment 1`] = `
"[
  {
    \\"type\\": 4,
    \\"data\\": {
      \\"href\\": \\"about:blank\\",
      \\"width\\": 1920,
      \\"height\\": 1080
    }
  },
  {
    \\"type\\": 2,
    \\"data\\": {
      \\"node\\": {
        \\"type\\": 0,
        \\"childNodes\\": [
          {
            \\"type\\": 1,
            \\"name\\": \\"html\\",
            \\"publicId\\": \\"\\",
            \\"systemId\\": \\"\\",
            \\"id\\": 2
          },
          {
            \\"type\\": 2,
            \\"tagName\\": \\"html\\",
            \\"attributes\\": {},
            \\"childNodes\\": [
              {
                \\"type\\": 2,
                \\"tagName\\": \\"head\\",
                \\"attributes\\": {},
                \\"childNodes\\": [
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"script\\",
                    \\"attributes\\": {
                      \\"type\\": \\"text/javascript\\"
                    },
                    \\"childNodes\\": [
                      {
                        \\"type\\": 3,
                        \\"textContent\\": \\"SCRIPT_PLACEHOLDER\\",
                        \\"id\\": 6
                      }
                    ],
                    \\"id\\": 5
                  }
                ],
                \\"id\\": 4
              },
              {
                \\"type\\": 2,
                \\"tagName\\": \\"body\\",
                \\"attributes\\": {},
                \\"childNodes\\": [
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n  \\\\n\\\\n\\",
                    \\"id\\": 8
                  }
                ],
                \\"id\\": 7
              }
            ],
            \\"id\\": 3
          }
        ],
        \\"id\\": 1
      },
      \\"initialOffset\\": {
        \\"left\\": 0,
        \\"top\\": 0
      }
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 0,
      \\"texts\\": [],
      \\"attributes\\": [],
      \\"removes\\": [],
      \\"adds\\": [
        {
          \\"parentId\\": 7,
          \\"nextId\\": null,
          \\"node\\": {
            \\"type\\": 2,
            \\"tagName\\": \\"div\\",
            \\"attributes\\": {},
            \\"childNodes\\": [],
            \\"id\\": 9
          }
        },
        {
          \\"parentId\\": 9,
          \\"nextId\\": null,
          \\"node\\": {
            \\"type\\": 2,
            \\"tagName\\": \\"div\\",
            \\"attributes\\": {},
            \\"childNodes\\": [],
            \\"id\\": 10
          }
        },
        {
          \\"parentId\\": 9,
          \\"nextId\\": 10,
          \\"node\\": {
            \\"type\\": 2,
            \\"tagName\\": \\"div\\",
            \\"attributes\\": {},
            \\"childNodes\\": [],
            \\"id\\": 11
          }
        }
      ]
    }
  }
]"
`;

exports[`mutation > ignored starting comment 1`] = `
"[
  {
    \\"type\\": 4,
    \\"data\\": {
      \\"href\\": \\"about:blank\\",
      \\"width\\": 1920,
      \\"height\\": 1080
    }
  },
  {
    \\"type\\": 2,
    \\"data\\": {
      \\"node\\": {
        \\"type\\": 0,
        \\"childNodes\\": [
          {
            \\"type\\": 1,
            \\"name\\": \\"html\\",
            \\"publicId\\": \\"\\",
            \\"systemId\\": \\"\\",
            \\"id\\": 2
          },
          {
            \\"type\\": 2,
            \\"tagName\\": \\"html\\",
            \\"attributes\\": {},
            \\"childNodes\\": [
              {
                \\"type\\": 2,
                \\"tagName\\": \\"head\\",
                \\"attributes\\": {},
                \\"childNodes\\": [
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"script\\",
                    \\"attributes\\": {
                      \\"type\\": \\"text/javascript\\"
                    },
                    \\"childNodes\\": [
                      {
                        \\"type\\": 3,
                        \\"textContent\\": \\"SCRIPT_PLACEHOLDER\\",
                        \\"id\\": 6
                      }
                    ],
                    \\"id\\": 5
                  }
                ],
                \\"id\\": 4
              },
              {
                \\"type\\": 2,
                \\"tagName\\": \\"body\\",
                \\"attributes\\": {},
                \\"childNodes\\": [
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n  \\\\n\\\\n\\",
                    \\"id\\": 8
                  }
                ],
                \\"id\\": 7
              }
            ],
            \\"id\\": 3
          }
        ],
        \\"id\\": 1
      },
      \\"initialOffset\\": {
        \\"left\\": 0,
        \\"top\\": 0
      }
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 0,
      \\"texts\\": [],
      \\"attributes\\": [],
      \\"removes\\": [],
      \\"adds\\": [
        {
          \\"parentId\\": 7,
          \\"nextId\\": null,
          \\"node\\": {
            \\"type\\": 2,
            \\"tagName\\": \\"div\\",
            \\"attributes\\": {},
            \\"childNodes\\": [],
            \\"id\\": 9
          }
        },
        {
          \\"parentId\\": 9,
          \\"nextId\\": null,
          \\"node\\": {
            \\"type\\": 2,
            \\"tagName\\": \\"div\\",
            \\"attributes\\": {},
            \\"childNodes\\": [],
            \\"id\\": 10
          }
        }
      ]
    }
  }
]"
`;
